src/Core/Search/Builder/TypeFiltersBuilder.php

```diff
@@ -0,0 +1,146 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Search\Builder;
+
+use PrestaShop\PrestaShop\Core\Search\Builder\TypedBuilder\TypedFiltersBuilderInterface;
+use PrestaShop\PrestaShop\Core\Search\Filters;
+
+/**
+ * Class TypeFiltersBuilder is an orchestrator which decided which builder is going to built the strongly
+ * typed Filter based on the defined filter class. It loops through a list of typed builders checking if
+ * they support the request class and uses the first it finds when a compatibility is found.
+ *
+ * If no specific type builder is found then the default builder is used.
+ */
+class TypeFiltersBuilder extends AbstractFiltersBuilder
+{
+    /**
+     * @var FiltersBuilderInterface
+     */
+    private $defaultBuilder;
+
+    /**
+     * @var TypedFiltersBuilderInterface[]
+     */
+    private $typedBuilders;
+
+    /** @var string */
+    private $filtersClass;
+
+    /**
+     * @var array|null
+     */
+    private $savedConfig = null;
+
+    /**
+     * @param FiltersBuilderInterface $defaultBuilder
+     * @param iterable|TypedFiltersBuilderInterface[]|null $typedBuilders
+     */
+    public function __construct(
+        FiltersBuilderInterface $defaultBuilder,
+        ?iterable $typedBuilders = null
+    ) {
+        $this->defaultBuilder = $defaultBuilder;
+        $this->typedBuilders = [];
+        if (!empty($typedBuilders)) {
+            foreach ($typedBuilders as $typedBuilder) {
+                $this->addTypedBuilder($typedBuilder);
+            }
+        }
+    }
+
+    /**
+     * @param TypedFiltersBuilderInterface $typedFiltersBuilder
+     *
+     * @return self
+     */
+    public function addTypedBuilder(TypedFiltersBuilderInterface $typedFiltersBuilder): self
+    {
+        $this->typedBuilders[] = $typedFiltersBuilder;
+        if (null !== $this->savedConfig) {
+            $typedFiltersBuilder->setConfig($this->savedConfig);
+        }
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setConfig(array $config)
+    {
+        $this->savedConfig = $config;
+
+        if (isset($config['filters_class'])) {
+            $this->filtersClass = $config['filters_class'];
+        }
+
+        $this->defaultBuilder->setConfig($config);
+        foreach ($this->typedBuilders as $typedBuilder) {
+            $typedBuilder->setConfig($config);
+        }
+
+        return parent::setConfig($config);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function buildFilters(Filters $filters = null)
```

Comment:
```suggestion
    public function buildFilters(Filters $filters = null): ?TypedFiltersBuilderInterface
```

Reply:
On this one not at all though!! If wee strong type it's with `Filters`

Reply:
same as above. (implements interface, need to respect its contract)